tapir/shifts/management/commands/apply_shift_cycle_start.py

```diff
@@ -0,0 +1,45 @@
+import datetime
+
+from django.core.management.base import BaseCommand
+from django.db.models import Max
+
+from tapir.shifts.models import ShiftCycleEntry, ShiftTemplateGroup, Shift
+from tapir.utils.shortcuts import get_monday
+
+
+class Command(BaseCommand):
+    help = "If a new cycle has started, remove one shift point from all active members."
+
+    def handle(self, *args, **options):
+        if ShiftCycleEntry.objects.exists():
+            last_cycle_date: datetime.date = ShiftCycleEntry.objects.aggregate(
+                Max("cycle_start_date")
+            )["cycle_start_date__max"]
+            new_cycle_start_date = last_cycle_date + datetime.timedelta(
+                days=ShiftCycleEntry.SHIFT_CYCLE_DURATION
+            )
+        else:
+            new_cycle_start_date = self.get_first_cycle_start_date()
+
+        if new_cycle_start_date is None:
+            return
+
+        while datetime.date.today() >= new_cycle_start_date:
+            ShiftCycleEntry.apply_cycle_start(new_cycle_start_date)
```

Comment:
From the name `apply_cycle_start`, it's not very obvious what's happening here (objects being created in the database, etc).

I said it back then, but I still feel this mechanism should have been a bit more generic. What we're doing is a "global account adjustment" or something like that, and we just want to make sure that we caught each user. It's easy to think of other mass operations that don't fit the cycle model: Reset everyone to zero (because of a general amnesty), clearing mass buffers (to make sure that people continously participate) or giving people shift credits for christmas because we can afford it. A shift cycle is just a very small abstraction on that, but right now it's baked pretty deep into the system.

Anyway, enough rambling, separating it out into `get_or_create_for_cycle_start` and `apply` or something like that would be cleaner, just to get what is happening.

Reply:
I'm not sure what parts of `apply_cycle_start` could be shared with the example operations you described?
Except iterating over users, what happens or when it happens depends on the operation, so I don't see what can be made generic.

I also don't understand what you mean with the split? Is it doing `get_credit_requirement_for_cycle `and creating the `ShiftAccountEntry` as an `apply` method in ShiftCycleEntry?